### PR TITLE
tons of functionality

### PR DESCRIPTION
**Pull request by SlimeQ**
I've been fleshing out this library for a personal project for a few months and at this point I've added a fairly long list of functionality:

- Pulling history
- Making edits
- Sending typing signals
- Uploading images
- Downloading images
- Getting user profile (pfp and nick)
- Getting room names from ID

The following message types now parse properly:
- EditEvent
- ImageMessageEvent
- RedactionEvent
- ReactionEvent

BaseRoomEvents now include EventId and Timestamp.

I only today got this into a semi-acceptable state to make a pull request. Some of it was written by ChatGPT and I rebelled against the established code style several times while adding code to areas I didn't understand. I've attempted to clean it up to a reasonable degree to fit into codebase, but I'm sure there are plenty of weird things throughout and it should be reviewed appropriately. For what it's worth, this code is currently running in production and seemingly all works properly.

### CODE REVIEW

**Comment by m-kus** [comment]
Hey! This is amazing, thank you so much for the PR!
We will also test it in our workflow to make sure everything works as expected

**Comment by SlimeQ** [comment]
:pray: 

https://github.com/baking-bad/matrix-dotnet-sdk/pull/11/files#diff-8217a0e68eb87c2a384ccdd80065834abfcad13f507adb70e6357e4fe83270e6

This spot in particular should probably be massaged a bit. I wasn't sure at the time how to get a reference to the event factory, so I made a static one. Personally I would go with a static class to house those factory functions if this were my own project, but idk if there are any unforseen implications to keeping that in the static context.

**Comment by SlimeQ** [comment]
I just went through and cleaned everything up and made it run. Downgraded the syntax to C# 9

Among some minor housekeeping changes:

- JoinedRoom/LeftRoom are now simply RoomResponse objects. These classes were identical and having both of them meant that code had to be copied twice in several places.
- I got rid of the MatrixRoom/RoomEvent factories; they did not seem to have any reason not to be static, and caused a ton of confusion when I was initially digging through this. Now those functions are static and contained by the object they are intended to create.
- MarkDig is now a dependency; this was in last night but not set up properly in the PR. I'm using this to generate the formatted_body on outgoing messages.


I cannot actually test this properly, as dotnet 5 is no longer supported on linux. However it does actually compile now.

**Comment by DevTown** [comment]
Will this PR be merged soon or is this project dead ?

**Comment by SlimeQ** [comment]
> Will this PR be merged soon or is this project dead ?

No idea, feel free to use my fork. I technically have some more stuff now around membership events that I haven't merged yet

**Comment by k-karuna** [comment]
Hi @DevTown @SlimeQ 
Sorry for long response, i will check out and test for bugs @SlimeQ changes in the coming days and merge them if everything well.

File: Matrix.Sdk/Matrix.Sdk.csproj
Comment (k-karuna) [inline Matrix.Sdk/Matrix.Sdk.csproj:1]: Please update Newtonsoft to 13.0.3

**Review by k-karuna — CHANGES_REQUESTED** [review]
@SlimeQ
`Matrix.Sdk/MatrixClientServiceExtensions.cs` should contain adding `services.AddSingleton<MediaService>();` otherwise it will not work in case of DependencyInjection builder.

Also please update Newtonsoft to 13.0.3

**Pull request merged**